rfcs/20251004-rms-norm/README.md

```diff
@@ -0,0 +1,161 @@
+This article provides more details regarding the method and showcases some experiments
+using frameworks like TensorFlow and PyTorch.
+
+Additionally, consider [Keras RMS Normalization Layer](https://github.com/keras-team/keras/blob/v3.9.2/keras/src/layers/normalization/rms_normalization.py)
+and [PyTorch RMSNorm](https://pytorch.org/docs/stable/generated/torch.nn.RMSNorm.html) APIs.
+
+## Proposal
+
+The proposal is to extend the LayerNorm primitive to support RMSNorm
+as this is a subset of existing functionality computations-wise.
+As for other Normalization primitives, the proposal is to not extend them at this point.
+
+For the work-in-progress POC, refer to [PR 3068](https://github.com/uxlfoundation/oneDNN/pull/3068).
+For more details about the proposal, see the following sections.
+
+Note, that previously, if users wanted to implement RMSNorm, they had to rely on the `dnnl_use_global_stats` flag,
+manually set the mean to zero, and compute the `RMS^2` statistics themselves.
+This approach is not convenient and adds extra complexity (computations, memory overhead) for the users.
+
+### Extension to the Normalization Flags
+
+This support of RMSNorm be achieved by adding a new flag to the existing `dnnl_normalization_flags_t`
+enum and `normalization_flags` enum in the C++ API:
+```cpp
+/// Flags for normalization primitives.
+enum class normalization_flags : unsigned {
+  <old flags...>
+
+  /// Use Root Mean Square (RMS) Layer Normalization.
+  /// In forward propagation, this means that the mean is set to zero, and RMS
+  /// is used instead of variance.
+  /// The RMS norm is provided as an output during forward propagation for
+  /// training.
+  /// In backward propagation, the library computes the derivative with respect
+  /// to the RMS norm, assuming that the mean is zero.
+  rms_norm = dnnl_rms_norm,
+};
+```
+
+```c
+/// Flags for normalization primitives.
+typedef enum {
+  <old flags...>
+
+  /// Use RMS normalization
+  ///
+  /// If specified:
+  ///  - on forward propagation use RMS normalization instead of true mean and
+  ///    variance
+  ///  - on backward propagation assume mean as zero
+  dnnl_rms_norm = 0x20U,
+} dnnl_normalization_flags_t;
+```
+
+The new flag `dnnl_rms_norm` will be used to indicate that the RMS
+normalization is used instead of the standard Layer Normalization.
+
+**Naming Option 2**:
+Alternative namings for the flag could signify that the mean is **assumed** to
+be zero, for instance `dnnl_{use_zero, no}_mean`. But this seems misleading
+as the mean is not explicitly set to zero, but simply ignored. Also it doesn't
+explciitly indicate what is happening with the variance.
+
+#### Compatibility with Existing Flags
+
+The combination of `dnnl_use_global_stats` and `dnnl_rms_norm` would be supported,
+but users must only provide pre-computed RMS norm statistics
+(instead of both mean and variance as required for Layer normalization).
```

Comment:
Could you precise that user need to pass pre-computed RMS norm with DNNL_ARG_VARIANCE.

Reply:
done in [92e9f5d](https://github.com/uxlfoundation/oneDNN/pull/3147/commits/92e9f5d29441fb429903c45d479499c2b2ce1cbf)